syz-verifier/monitoring_api.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2021 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"encoding/json"
+	"net/http"
+	"time"
+)
+
+// Monitor provides http based data for the syz-verifier monitoring.
+type Monitor struct {
+	externalStats *Stats
+}
+
+// MakeMonitor creates the Monitor instance.
+func MakeMonitor() *Monitor {
+	return &Monitor{}
+}
+
+// InitHTTPHandlers initializes the API routing.
+func (monitor *Monitor) InitHTTPHandlers() {
+	http.Handle("/api/stats.json", &statsHandler{
+		Src: monitor.externalStats,
+	})
+
+	http.HandleFunc("/", func(writer http.ResponseWriter, request *http.Request) {
+		writer.Write([]byte("<a href='api/stats.json'>stats_json</a>"))
+	})
+}
+
+// SetStatsTracking points Monitor to the Stats object to report.
+func (monitor *Monitor) SetStatsTracking(s *Stats) {
+	monitor.externalStats = s
+}
+
+// ListenAndServe starts the server.
+func (monitor *Monitor) ListenAndServe(addr string) error {
+	return http.ListenAndServe(addr, nil)
+}
+
+type statsHandler struct {
+	Src *Stats
+}
+
+// statsJSON provides information for the "/api/stats.json" render.
+type statsJSON struct {
+	StartTime        time.Time
+	TotalMismatches  int
+	TotalProgs       int
+	FlakyProgs       int
+	MismatchingProgs int
+	AverExecSpeed    int
+}
+
+// ServeHTTP render the statsHandler.Stats object.
+func (h *statsHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	setJSONAPIHeaders(w)
+
+	data, err := json.MarshalIndent(
+		&statsJSON{
+			StartTime:        h.Src.StartTime,
+			TotalMismatches:  h.Src.TotalMismatches,
+			TotalProgs:       h.Src.TotalProgs,
+			FlakyProgs:       h.Src.FlakyProgs,
+			MismatchingProgs: h.Src.MismatchingProgs,
+			AverExecSpeed:    60 * h.Src.TotalProgs / int(1+time.Since(h.Src.StartTime).Seconds()),
+		},
+		"",
+		"\t",
+	)
+	if err != nil {
+		return
```

Comment:
Use https://pkg.go.dev/net/http#Error here to notify the remote end.

Reply:
Good one. Thank you.